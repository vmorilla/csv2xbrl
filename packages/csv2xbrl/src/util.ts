import { Writable } from "stream";

export class StringWritable extends Writable {
    private _data = '';

    _write(chunk: any, encoding: BufferEncoding, callback: (error?: Error | null) => void): void {
        this._data += chunk.toString();
        callback();
    }

    toString(): string {
        return this._data;
    }
}
