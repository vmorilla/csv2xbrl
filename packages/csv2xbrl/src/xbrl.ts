
export const namespaces = {
    'xsd': 'http://www.w3.org/2001/XMLSchema',
    'xsi': 'http://www.w3.org/2001/XMLSchema-instance',
    'xbrli': 'http://www.xbrl.org/2003/instance',
    'link': 'http://www.xbrl.org/2003/linkbase',
    'xlink': 'http://www.w3.org/1999/xlink',
    'xbrldi': 'http://xbrl.org/2006/xbrldi',
    'xbrldt': 'http://xbrl.org/2005/xbrldt',
    'tp': 'http://xbrl.org/2016/taxonomy-package'
};


