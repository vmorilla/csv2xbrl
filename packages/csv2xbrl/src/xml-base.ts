import xpath from "xpath";
import { resolve } from "url";
import { DOMParser } from "xmldom";

/**
 * Resolves a URL in the context of an XML node, considering the semantics of xml:base attributes.
 * @param url URL to resolve
 * @param node Node to resolve URL against
 * @returns resolved URL
 */
export function resolveNodeUrl(url: string, node: Node): string {
    const select = xpath.useNamespaces({ 'xml': 'http://www.w3.org/XML/1998/namespace' });
    let baseUri = (node.ownerDocument || node).baseURI;
    const xml_base_attrs = select('ancestor-or-self::*/@xml:base', node) as Attr[];
    for (const xml_base_attr of xml_base_attrs) {
        baseUri = baseUri ? resolve(baseUri, xml_base_attr.value) : xml_base_attr.value;
    }

    return baseUri ? resolve(baseUri, url) : url;
}

/**
 * Parses a document from a string, setting the baseURI to the provided URL
 * The implementation overrides a read-only attribute at document level
 * @param xmlString 
 * @param url 
 * @returns 
 */
export function parseXMLFromString(xmlString: string, url?: string): Document {
    const doc = new DOMParser().parseFromString(xmlString, "application/xml");
    if (url)
        (doc as any).baseURI = url;
    return doc;
}
