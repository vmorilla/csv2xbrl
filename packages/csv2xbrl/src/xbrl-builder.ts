import { namespaces } from "./xbrl";
import { XMLWriter } from "./simple-xml-writer";

const { xsi, xbrli, link, xlink, xbrldi } = namespaces;
const instance_namespaces = {
    '': xbrli, xsi, link, xlink, xbrldi
}

// Dimension sets do not include OIM's concept dimensions
export type FullDimensions = { unit?: Unit } & ContextDimensions;
export type ContextDimensions = { entity: Entity, period: Period } & ETDimensions;
export type ETDimensions = { [key: string]: ExplicitMember | TypedMember | undefined };

export type ExplicitMember = string | undefined; // OIM allows null value for the representation of the default member of a dimensions
export type TypedMember = [string, string | undefined];
export type Decimals = number | "INF" | undefined;
export type Period = string | [string, string];
export type Unit = string | undefined;
export type Entity = [string, string];

/**
 * The XBRL builder does not make checks on the consistency of the usage of prefixes. It assumes user prefixes
 * (the ones used for dimension names and dimension values) are included in the constructor.
 */
export class XbrlBuilder {
    private xbrl: XMLWriter;

    // Map standard prefixes to prefixes valid in the instance document
    // Necessary to consider the case where a standard prefix is used for a different namespace in the instance document
    private prefix: Record<string, (arg: string) => string> = {};

    // Map of contexts from their surrogated key to its id
    private _lastContextId = 0;
    private _contexts: Map<string, string> = new Map();

    // Map of units from their surrogated key to its id
    private _lastUnitId = 0;
    private _unit: Map<string, string> = new Map();


    constructor(entryPoints: string[], ns_prefixes: { [prefix: string]: string }, output: NodeJS.WritableStream) {

        // Initializes the writer
        this.xbrl = new XMLWriter(output, { version: "1.0", encoding: "UTF-8" }, ns_prefixes);

        // Register xbrl instance prefixes
        for (const [prefix, ns] of Object.entries(instance_namespaces)) {
            const newPrefix = this.xbrl.registerNamespace(prefix, ns);
            this.prefix[prefix] = prefixer(newPrefix);
        }

        this.xbrl.elm("xbrl");
        const link = this.prefix.link;
        const xlink = this.prefix.xlink;

        for (const entryPoint of entryPoints) {
            if (entryPoint.trim().toLowerCase().endsWith('.xsd')) {
                this.xbrl.felm(link('schemaRef'), { [xlink('href')]: entryPoint });
            } else {
                this.xbrl.felm(link('linkbaseRef'), { [xlink('href')]: entryPoint });
            }
        }

        return this;
    }

    public context(dimensions: ContextDimensions): string {

        const key = this.surrogatedKey(dimensions);
        if (this._contexts.has(key)) {
            return this._contexts.get(key)!;
        }
        else {
            const id = `c${++this._lastContextId}`;
            this._contexts.set(key, id);
            this.addContext(id, dimensions);
            return id;
        }
    }

    private addContext(id: string, contextDimensions: ContextDimensions) {
        const { entity: [entitySchema, entityIdentifier], period, ...restDimensions } = contextDimensions;
        this.xbrl.elm('context', { id });
        this.xbrl.elm('entity').felm('identifier', { scheme: entitySchema }, entityIdentifier).end();
        if (period instanceof Array) {
            this.xbrl.elm('period')
                .felm('startDate', {}, period[0])
                .felm('endDate', {}, period[1])
                .end();
        }
        else {
            this.xbrl.elm('period').felm('instant', {}, period).end();
        }
        const xbrldi = this.prefix.xbrldi;
        if (Object.keys(restDimensions).length > 0) {
            this.xbrl.elm('scenario');
            for (const [key, value] of Object.entries(restDimensions)) {
                if (typeof value === 'string') {
                    this.xbrl.felm(xbrldi("explicitMember"), { dimension: key }, value);
                } else {
                    if (typeof value !== 'undefined') {
                        const [domain, domainValue] = value as TypedMember;
                        if (typeof domainValue !== 'undefined') {
                            this.xbrl.elm(xbrldi("typedMember"), { dimension: key }).felm(domain, {}, domainValue).end();
                        }
                    }
                }
            }
            this.xbrl.end(); // Scenario
        }
        this.xbrl.end(); // Context
    }

    private surrogatedKey(dimensions: ContextDimensions) {
        const orderedEntries = Object.entries(dimensions).sort(([keyA], [keyB]) => keyA.localeCompare(keyB));
        return JSON.stringify(orderedEntries);
    }

    public unit(unit: string) {
        if (this._contexts.has(unit)) {
            return this._contexts.get(unit)!;
        }
        else {
            const id = `u${++this._lastUnitId}`;
            this._contexts.set(unit, id);
            this.xbrl.elm('unit', { id }).felm('measure', {}, unit).end();
            return id;
        }
    }

    public numericFact(concept: string, decimals: Decimals, unit: string, contextDimensions: ContextDimensions, value: number) {
        const contextRef = this.context(contextDimensions);
        const unitRef = this.unit(unit!);
        this.xbrl.felm(concept!, { decimals: (decimals ?? "INF").toString(), contextRef, unitRef }, value.toString());
        return this;
    }

    public nonNumericFact(concept: string, contextDimensions: ContextDimensions, value: string) {
        const contextRef = this.context(contextDimensions);
        this.xbrl.felm(concept!, { contextRef }, value);
        return this;
    }

    public comment(text: string) {
        this.xbrl.comment(text);

        return this;
    }

    public end() {
        this.xbrl.end();
    }
}

/**
 * Returns a function that adds a prefix to a localName to create a QName
 * @param prefix 
 * @returns 
 */
function prefixer(prefix: string) {
    return (name: string) => `${prefix}:${name}`;
}

