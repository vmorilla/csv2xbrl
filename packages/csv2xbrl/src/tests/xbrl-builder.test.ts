
import { StringWritable } from '../util';
import { namespaces } from '../xbrl';
import { XbrlBuilder } from '../xbrl-builder';
import { parseXMLFromString } from '../xml-base';
import xpath from 'xpath';

const select = xpath.useNamespaces(namespaces);

describe(XbrlBuilder, () => {
    it("should create a basic instance with no facts", () => {
        const output = new StringWritable();
        const builder = new XbrlBuilder([], {}, output).end();
        expect(output.toString()).toMatchSnapshot();
    });

    it("should create a basic instance with target schmea but no facts", () => {
        const output = new StringWritable();
        const builder = new XbrlBuilder(["https://www.acme.org/taxonomy.xsd"], {}, output).end();
        expect(output.toString()).toMatchSnapshot();
    });

    it("uses a different prefix if there is a conflict with user defined ones", () => {
        const output = new StringWritable();
        const builder = new XbrlBuilder([], { xbrldi: "http://acme.org/xbrldi" }, output)
            .nonNumericFact("xbrldi:concept", { entity: ["http://iso.com/lei", "FAKELEI"], period: "2020-01-01", "xbrldi:dim": "xbrldi:exp" }, "Hello World")
            .end();
        expect(output.toString()).toMatchSnapshot();
        const xbrl = parseXMLFromString(output.toString());
        const dim = select("//xbrldi:explicitMember", xbrl, true) as Node;
        const [prefix, _] = dim.nodeName.split(":");
        expect(xbrl.documentElement.getAttributeNS("http://www.w3.org/2000/xmlns/", prefix)).toBe("http://xbrl.org/2006/xbrldi");
    });

    it("adds a nonNumeric fact with a simple context of instant type", () => {
        const output = new StringWritable();
        const builder = new XbrlBuilder([], { acme: "acme.org/taxonomy" }, output)
            .nonNumericFact("acme:Sales", { entity: ["http://iso.com/lei", "FAKELEI"], period: "2020-01-01" }, "Hello World")
            .end();
        expect(output.toString()).toMatchSnapshot();
        const xbrl = parseXMLFromString(output.toString());
        expect((select("//xbrli:instant", xbrl) as Node[]).length).toBe(1);
        expect(output.toString()).toMatchSnapshot();

    });

    it("adds a nonNumeric fact with a simple context of duration type", () => {
        const output = new StringWritable();
        const builder = new XbrlBuilder([], { acme: "acme.org/taxonomy" }, output)
            .nonNumericFact("acme:Sales", { entity: ["http://iso.com/lei", "FAKELEI"], period: ["2020-01-01", "2020-12-31"] }, "Hello World")
            .end();
        expect(output.toString()).toMatchSnapshot();
        const xbrl = parseXMLFromString(output.toString());
        expect((select("//xbrli:startDate", xbrl) as Node[]).length).toBe(1);
        expect((select("//xbrli:endDate", xbrl) as Node[]).length).toBe(1);
        expect(output.toString()).toMatchSnapshot();
    });

    it("adds a fact with a typed dimension", () => {
        const output = new StringWritable();
        const builder = new XbrlBuilder([], { acme: "acme.org/taxonomy" }, output)
            .nonNumericFact("acme:Sales", { entity: ["http://iso.com/lei", "FAKELEI"], period: "2020-01-01", "acme:dim": ["acme:LE", "TyuepedVal"] }, "Hello World")
            .end();
        expect(output.toString()).toMatchSnapshot();
        const xbrl = parseXMLFromString(output.toString());
        expect((select("//xbrldi:typedMember", xbrl) as Node[]).length).toBe(1);
    });

    it("adds a fact with an explicit dimension dimension", () => {
        const output = new StringWritable();
        const builder = new XbrlBuilder([], { acme: "acme.org/taxonomy" }, output)
            .nonNumericFact("acme:Sales", { entity: ["http://iso.com/lei", "FAKELEI"], period: "2020-01-01", "acme:dim": "acme:explicit" }, "Hello World")
            .end();
        expect(output.toString()).toMatchSnapshot();
        const xbrl = parseXMLFromString(output.toString());
        expect((select("//xbrldi:explicitMember", xbrl) as Node[]).length).toBe(1);
    });

    it("adds a numeric fact with a simple context of instant type", () => {
        const output = new StringWritable();
        const builder = new XbrlBuilder([], { acme: "acme.org/taxonomy" }, output)
            .numericFact("acme:Sales", 2, "iso4217:USD", { entity: ["http://iso.com/lei", "FAKELEI"], period: "2020-01-01" }, 100)
            .end();
        expect(output.toString()).toMatchSnapshot();
        const xbrl = parseXMLFromString(output.toString());
        expect((select("//xbrli:instant", xbrl) as Node[]).length).toBe(1);
    });


});
