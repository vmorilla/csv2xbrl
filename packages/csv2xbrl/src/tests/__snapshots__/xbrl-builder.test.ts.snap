// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`XbrlBuilder adds a fact with a typed dimension 1`] = `
"<?xml version="1.0" encoding="UTF-8"?>
<xbrl
	xmlns="http://www.xbrl.org/2003/instance"
	xmlns:acme="acme.org/taxonomy"
	xmlns:link="http://www.xbrl.org/2003/linkbase"
	xmlns:xbrldi="http://xbrl.org/2006/xbrldi"
	xmlns:xlink="http://www.w3.org/1999/xlink"
	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
	<context id="c1">
		<entity>
			<identifier scheme="http://iso.com/lei">FAKELEI</identifier>
		</entity>
		<period>
			<instant>2020-01-01</instant>
		</period>
		<scenario>
			<xbrldi:typedMember dimension="acme:dim">
				<acme:LE>TyuepedVal</acme:LE>
			</xbrldi:typedMember>
		</scenario>
	</context>
	<acme:Sales contextRef="c1">Hello World</acme:Sales>
</xbrl>
"
`;

exports[`XbrlBuilder adds a fact with an explicit dimension dimension 1`] = `
"<?xml version="1.0" encoding="UTF-8"?>
<xbrl
	xmlns="http://www.xbrl.org/2003/instance"
	xmlns:acme="acme.org/taxonomy"
	xmlns:link="http://www.xbrl.org/2003/linkbase"
	xmlns:xbrldi="http://xbrl.org/2006/xbrldi"
	xmlns:xlink="http://www.w3.org/1999/xlink"
	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
	<context id="c1">
		<entity>
			<identifier scheme="http://iso.com/lei">FAKELEI</identifier>
		</entity>
		<period>
			<instant>2020-01-01</instant>
		</period>
		<scenario>
			<xbrldi:explicitMember dimension="acme:dim">acme:explicit</xbrldi:explicitMember>
		</scenario>
	</context>
	<acme:Sales contextRef="c1">Hello World</acme:Sales>
</xbrl>
"
`;

exports[`XbrlBuilder adds a nonNumeric fact with a simple context of duration type 1`] = `
"<?xml version="1.0" encoding="UTF-8"?>
<xbrl
	xmlns="http://www.xbrl.org/2003/instance"
	xmlns:acme="acme.org/taxonomy"
	xmlns:link="http://www.xbrl.org/2003/linkbase"
	xmlns:xbrldi="http://xbrl.org/2006/xbrldi"
	xmlns:xlink="http://www.w3.org/1999/xlink"
	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
	<context id="c1">
		<entity>
			<identifier scheme="http://iso.com/lei">FAKELEI</identifier>
		</entity>
		<period>
			<startDate>2020-01-01</startDate>
			<endDate>2020-12-31</endDate>
		</period>
	</context>
	<acme:Sales contextRef="c1">Hello World</acme:Sales>
</xbrl>
"
`;

exports[`XbrlBuilder adds a nonNumeric fact with a simple context of duration type 2`] = `
"<?xml version="1.0" encoding="UTF-8"?>
<xbrl
	xmlns="http://www.xbrl.org/2003/instance"
	xmlns:acme="acme.org/taxonomy"
	xmlns:link="http://www.xbrl.org/2003/linkbase"
	xmlns:xbrldi="http://xbrl.org/2006/xbrldi"
	xmlns:xlink="http://www.w3.org/1999/xlink"
	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
	<context id="c1">
		<entity>
			<identifier scheme="http://iso.com/lei">FAKELEI</identifier>
		</entity>
		<period>
			<startDate>2020-01-01</startDate>
			<endDate>2020-12-31</endDate>
		</period>
	</context>
	<acme:Sales contextRef="c1">Hello World</acme:Sales>
</xbrl>
"
`;

exports[`XbrlBuilder adds a nonNumeric fact with a simple context of instant type 1`] = `
"<?xml version="1.0" encoding="UTF-8"?>
<xbrl
	xmlns="http://www.xbrl.org/2003/instance"
	xmlns:acme="acme.org/taxonomy"
	xmlns:link="http://www.xbrl.org/2003/linkbase"
	xmlns:xbrldi="http://xbrl.org/2006/xbrldi"
	xmlns:xlink="http://www.w3.org/1999/xlink"
	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
	<context id="c1">
		<entity>
			<identifier scheme="http://iso.com/lei">FAKELEI</identifier>
		</entity>
		<period>
			<instant>2020-01-01</instant>
		</period>
	</context>
	<acme:Sales contextRef="c1">Hello World</acme:Sales>
</xbrl>
"
`;

exports[`XbrlBuilder adds a nonNumeric fact with a simple context of instant type 2`] = `
"<?xml version="1.0" encoding="UTF-8"?>
<xbrl
	xmlns="http://www.xbrl.org/2003/instance"
	xmlns:acme="acme.org/taxonomy"
	xmlns:link="http://www.xbrl.org/2003/linkbase"
	xmlns:xbrldi="http://xbrl.org/2006/xbrldi"
	xmlns:xlink="http://www.w3.org/1999/xlink"
	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
	<context id="c1">
		<entity>
			<identifier scheme="http://iso.com/lei">FAKELEI</identifier>
		</entity>
		<period>
			<instant>2020-01-01</instant>
		</period>
	</context>
	<acme:Sales contextRef="c1">Hello World</acme:Sales>
</xbrl>
"
`;

exports[`XbrlBuilder adds a numeric fact with a simple context of instant type 1`] = `
"<?xml version="1.0" encoding="UTF-8"?>
<xbrl
	xmlns="http://www.xbrl.org/2003/instance"
	xmlns:acme="acme.org/taxonomy"
	xmlns:link="http://www.xbrl.org/2003/linkbase"
	xmlns:xbrldi="http://xbrl.org/2006/xbrldi"
	xmlns:xlink="http://www.w3.org/1999/xlink"
	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
	<context id="c1">
		<entity>
			<identifier scheme="http://iso.com/lei">FAKELEI</identifier>
		</entity>
		<period>
			<instant>2020-01-01</instant>
		</period>
	</context>
	<unit id="u1">
		<measure>iso4217:USD</measure>
	</unit>
	<acme:Sales contextRef="c1" decimals="2" unitRef="u1">100</acme:Sales>
</xbrl>
"
`;

exports[`XbrlBuilder should create a basic instance with no facts 1`] = `
"<?xml version="1.0" encoding="UTF-8"?>
<xbrl
	xmlns="http://www.xbrl.org/2003/instance"
	xmlns:link="http://www.xbrl.org/2003/linkbase"
	xmlns:xbrldi="http://xbrl.org/2006/xbrldi"
	xmlns:xlink="http://www.w3.org/1999/xlink"
	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
</xbrl>
"
`;

exports[`XbrlBuilder should create a basic instance with target schmea but no facts 1`] = `
"<?xml version="1.0" encoding="UTF-8"?>
<xbrl
	xmlns="http://www.xbrl.org/2003/instance"
	xmlns:link="http://www.xbrl.org/2003/linkbase"
	xmlns:xbrldi="http://xbrl.org/2006/xbrldi"
	xmlns:xlink="http://www.w3.org/1999/xlink"
	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
	<link:schemaRef xlink:href="https://www.acme.org/taxonomy.xsd" />
</xbrl>
"
`;

exports[`XbrlBuilder uses a different prefix if there is a conflict with user defined ones 1`] = `
"<?xml version="1.0" encoding="UTF-8"?>
<xbrl
	xmlns="http://www.xbrl.org/2003/instance"
	xmlns:link="http://www.xbrl.org/2003/linkbase"
	xmlns:xbrldi="http://acme.org/xbrldi"
	xmlns:xbrldi0="http://xbrl.org/2006/xbrldi"
	xmlns:xlink="http://www.w3.org/1999/xlink"
	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
	<context id="c1">
		<entity>
			<identifier scheme="http://iso.com/lei">FAKELEI</identifier>
		</entity>
		<period>
			<instant>2020-01-01</instant>
		</period>
		<scenario>
			<xbrldi0:explicitMember dimension="xbrldi:dim">xbrldi:exp</xbrldi0:explicitMember>
		</scenario>
	</context>
	<xbrldi:concept contextRef="c1">Hello World</xbrldi:concept>
</xbrl>
"
`;
